File: backend/questionnaires.py
"""
Mental Health Questionnaires & Scoring
========================================
Definitions and scoring logic for PHQ-9, GAD-7, WHO-5, and PSS.
"""

from typing import List, Dict, Any, Optional

class QuestionnaireDefinition:
    def __init__(self, id: str, name: str, description: str, questions: List[str], options: List[Dict[str, Any]]):
        self.id = id
        self.name = name
        self.description = description
        self.questions = questions
        self.options = options

    def score(self, answers: List[int]) -> dict:
        raise NotImplementedError("Subclasses must implement score()")


class PHQ9(QuestionnaireDefinition):
    def __init__(self):
        super().__init__(
            id="PHQ-9",
            name="PHQ-9: Questionário de Saúde do Paciente",
            description="Uma breve ferramenta de rastreio para gravidade da depressão.",
            questions=[
                "Você tem sentido menos vontade ou prazer em fazer as coisas que normalmente gosta?",
                "Você tem se sentido mais triste, desanimado(a) ou sem muita perspectiva ultimamente?",
                "Como tem sido o seu sono? Tem sido difícil dormir ou você tem sentido necessidade de dormir em excesso?",
                "Você tem se sentido muito cansado(a) ou notou que sua energia está mais baixa que o normal?",
                "Você notou mudanças no apetite, como sentir pouca fome ou acabar comendo bem mais que o costume?",
                "Você tem sido muito duro(a) consigo mesmo(a), sentindo-se um fracasso ou que decepcionou quem você ama?",
                "Tem sido difícil manter o foco em atividades simples, como ler uma notícia ou assistir a algo na TV?",
                "Você sentiu seu corpo mais lento para falar e agir, ou, ao contrário, uma agitação que não te deixa relaxar?",
                "Você chegou a ter pensamentos de que seria melhor não estar mais aqui ou de se machucar de alguma forma?"
            ],
            options=[
                {"text": "Nenhum dia", "value": 0},
                {"text": "Em alguns dias", "value": 1},
                {"text": "Na maioria dos dias", "value": 2},
                {"text": "Quase todos os dias", "value": 3}
            ]
        )

    def score(self, answers: List[int]) -> dict:
        total = sum(answers)
        risk = "low"
        if total >= 20: risk = "critical"
        elif total >= 15: risk = "high"
        elif total >= 10: risk = "moderate"
        
        return {"total_score": total, "risk_level": risk}


class GAD7(QuestionnaireDefinition):
    def __init__(self):
        super().__init__(
            id="GAD-7",
            name="GAD-7: Escala de Ansiedade Generalizada",
            description="Uma breve ferramenta de rastreio para gravidade da ansiedade.",
            questions=[
                "Tem sentido uma sensação de ansiedade e tensão constante?",
                "Tem sido difícil desligar os pensamentos ou controlar a quantidade de preocupações que lhe vêm à cabeça?",
                "Tem dado por si a preocupar-se excessivamente com vários assuntos diferentes do dia a dia ao mesmo tempo?",
                "Tem sentido dificuldade em conseguir relaxar ou em ter um momento de verdadeira calma para si?",
                "Tem sentido uma inquietação no corpo, daquelas em que se torna muito difícil conseguir estar quieto/a ou sossegado/a?",
                "Tem notado que anda com menos paciência do que o habitual, aborrecendo-se ou irritando-se com facilidade?",
                "Tem sentido um receio súbito ou uma sensação de alarme, como se algo de muito mau estivesse prestes a acontecer?"
            ],
            options=[
                {"text": "Nunca", "value": 0},
                {"text": "Em vários dias", "value": 1},
                {"text": "Em mais de metade dos dias", "value": 2},
                {"text": "Em quase todos os dias", "value": 3}
            ]
        )

    def score(self, answers: List[int]) -> dict:
        total = sum(answers)
        risk = "low"
        if total >= 15: risk = "high"
        elif total >= 10: risk = "moderate"
        elif total >= 5: risk = "low" # mild anxiety is low risk in our MVP
        
        return {"total_score": total, "risk_level": risk}


class WHO5(QuestionnaireDefinition):
    def __init__(self):
        super().__init__(
            id="WHO-5",
            name="WHO-5: Índice de Bem-Estar",
            description="Uma breve avaliação do bem-estar subjetivo atual.",
            questions=[
                "Tem-se sentido alegre, bem-disposto(a) e com bom astral ultimamente?",
                "Com que frequência se tem sentido calmo(a), tranquilo(a) e relaxado(a)?",
                "Tem-se sentido com energia, ativo(a) e com vigor para realizar as suas tarefas?",
                "Ao acordar, sente que teve um sono revigorante e que está realmente descansado(a)?",
                "Sente que o seu dia a dia tem sido preenchido com coisas e atividades que despertam o seu interesse?"
            ],
            options=[
                {"text": "Todo o tempo", "value": 5},
                {"text": "A maior parte do tempo", "value": 4},
                {"text": "Mais de metade do tempo", "value": 3},
                {"text": "Menos de metade do tempo", "value": 2},
                {"text": "Algumas vezes", "value": 1},
                {"text": "Nunca", "value": 0}
            ]
        )

    def score(self, answers: List[int]) -> dict:
        raw_total = sum(answers)
        percentage = raw_total * 4 # 0-100 scale
        
        risk = "low"
        if percentage <= 28:
            risk = "high" # Indicates likely depression
        elif percentage <= 50:
            risk = "moderate"
        
        return {"total_score": percentage, "raw_score": raw_total, "risk_level": risk}


class PSS(QuestionnaireDefinition):
    def __init__(self):
        super().__init__(
            id="PSS-10",
            name="PSS-10: Escala de Stress Percebido",
            description="Uma breve avaliação da perceção de stress nas últimas semanas.",
            questions=[
                "Você tem se sentido afetado(a) ou abalado(a) por coisas inesperadas que aconteceram recentemente?",
                "Com que frequência sentiu que não tinha controle sobre as coisas realmente importantes da sua vida?",
                "Você tem se sentido muito nervoso(a) ou 'sob pressão' ultimamente?",
                "Com que frequência sentiu confiança na sua própria capacidade de lidar com os seus problemas pessoais?",
                "Você tem sentido que as coisas estão a correr como gostaria ou planeou?",
                "Com que frequência sentiu que não conseguia dar conta de todas as tarefas e obrigações que tinha para fazer?",
                "Tem sido capaz de controlar as irritações ou pequenas chateações do dia a dia?",
                "Com que frequência sentiu que tinha tudo sob controlo e que as coisas estavam a caminhar bem?",
                "Tem se sentido furioso(a) ou muito irritado(a) por causa de coisas que fugiram totalmente do seu controlo?",
                "Com que frequência sentiu que as dificuldades se acumulavam tanto que já não as conseguia ultrapassar?"
            ],
            options=[
                {"text": "Nunca", "value": 0},
                {"text": "Quase nunca", "value": 1},
                {"text": "Às vezes", "value": 2},
                {"text": "Com frequência", "value": 3},
                {"text": "Com muita frequência", "value": 4}
            ]
        )

    def score(self, answers: List[int]) -> dict:
        # Reverse scoring for questions 4, 5, 7, and 8 (0-indexed: 3, 4, 6, 7)
        reversed_indices = [3, 4, 6, 7]
        total = 0
        for i, val in enumerate(answers):
            if i in reversed_indices:
                total += (4 - val)
            else:
                total += val

        risk = "low"
        if total >= 27: risk = "high"
        elif total >= 14: risk = "moderate"
        
        return {"total_score": total, "risk_level": risk}


# Registry for easy lookup
QUESTIONNAIRES = {
    "PHQ-9": PHQ9(),
    "GAD-7": GAD7(),
    "WHO-5": WHO5(),
    "PSS-10": PSS()
}

def get_questionnaire(q_id: str) -> Optional[QuestionnaireDefinition]:
    return QUESTIONNAIRES.get(q_id.upper())

def score_assessment(q_id: str, answers: List[int]) -> dict:
    q = get_questionnaire(q_id)
    if not q:
        raise ValueError(f"Unknown questionnaire ID: {q_id}")
    if len(answers) != len(q.questions):
        raise ValueError(f"Expected {len(q.questions)} answers, got {len(answers)}")
    
    return q.score(answers)

File: backend/src/data/questionnaires.js
const QUESTIONNAIRES = {
  'PHQ-9': {
    id: 'PHQ-9',
    en: {
      name: 'PHQ-9: Patient Health Questionnaire',
      description: 'A brief screening tool for depression severity.',
      questions: [
        'Have you felt little interest or pleasure in doing things that you normally enjoy?',
        'Have you been feeling down, depressed, or hopeless lately?',
        'How has your sleep been? Have you had difficulty sleeping or felt the need to sleep too much?',
        'Have you been feeling very tired or noticed that your energy is lower than usual?',
        'Have you noticed changes in your appetite, such as eating much less or much more than usual?',
        'Have you been hard on yourself, feeling like a failure or that you\'ve disappointed people you love?',
        'Have you had difficulty focusing on simple activities, like reading or watching TV?',
        'Have you noticed your body moving more slowly or speaking more slowly, or the opposite - feeling agitated and restless?',
        'Have you had thoughts that it would be better if you weren\'t here, or thoughts of hurting yourself?'
      ],
      options: [
        { text: 'Not at all', value: 0 },
        { text: 'Several days', value: 1 },
        { text: 'More than half the days', value: 2 },
        { text: 'Nearly every day', value: 3 }
      ]
    },
    pt: {
      name: 'PHQ-9: Questionário de Saúde do Paciente',
      description: 'Uma ferramenta breve de rastreio para a gravidade da depressão.',
      questions: [
        'Tem sentido pouco interesse ou prazer em fazer coisas que normalmente aprecia?',
        'Tem-se sentido em baixo, deprimido ou sem esperança ultimamente?',
        'Como tem estado o seu sono? Tem tido dificuldade em dormir ou sentido necessidade de dormir em excesso?',
        'Tem-se sentido muito cansado ou notado que a sua energia está mais baixa do que o normal?',
        'Notou alterações no apetite, comendo muito menos ou muito mais do que o habitual?',
        'Tem sido duro consigo mesmo, sentindo-se um fracasso ou que desiludiu as pessoas que ama?',
        'Tem tido dificuldade em concentrar-se em atividades simples, como ler ou ver televisão?',
        'Notou que o seu corpo se move ou fala mais devagar, ou o oposto - sentindo-se agitado e inquieto?',
        'Teve pensamentos de que seria melhor não estar aqui, ou pensamentos de se magoar?'
      ],
      options: [
        { text: 'Nenhum dia', value: 0 },
        { text: 'Vários dias', value: 1 },
        { text: 'Mais da metade dos dias', value: 2 },
        { text: 'Quase todos os dias', value: 3 }
      ]
    }
  },
  'GAD-7': {
    id: 'GAD-7',
    en: {
      name: 'GAD-7: Generalized Anxiety Disorder Scale',
      description: 'A brief screening tool for anxiety severity.',
      questions: [
        'Have you felt a constant sense of anxiety and tension?',
        'Have you found it difficult to turn off your thoughts or control how much worry comes to your mind?',
        'Have you been excessively worried about various different everyday issues at the same time?',
        'Have you had difficulty relaxing or having a moment of true calm for yourself?',
        'Have you felt restlessness in your body, making it very difficult to sit still or feel settled?',
        'Have you noticed that you\'ve been less patient than usual, getting annoyed or irritated easily?',
        'Have you experienced a sudden fear or a sense of alarm, as if something very bad was about to happen?'
      ],
      options: [
        { text: 'Never', value: 0 },
        { text: 'Several days', value: 1 },
        { text: 'More than half the days', value: 2 },
        { text: 'Nearly every day', value: 3 }
      ]
    },
    pt: {
      name: 'GAD-7: Escala de Ansiedade Generalizada',
      description: 'Uma ferramenta breve de rastreio para a gravidade da ansiedade.',
      questions: [
        'Tem sentido uma constante sensação de ansiedade e tensão?',
        'Tem tido dificuldade em desligar os pensamentos ou controlar a preocupação?',
        'Tem-se preocupado excessivamente com várias questões do dia-a-dia ao mesmo tempo?',
        'Tem tido dificuldade em relaxar ou ter um momento de verdadeira calma para si?',
        'Tem sentido inquietação no corpo, sendo muito difícil estar quieto ou sentir-se calmo?',
        'Notou que tem estado menos paciente do que o normal, irritando-se com facilidade?',
        'Experienciou um medo repentino ou um sentimento de alarme, como se algo muito mau fosse acontecer?'
      ],
      options: [
        { text: 'Nunca', value: 0 },
        { text: 'Vários dias', value: 1 },
        { text: 'Mais da metade dos dias', value: 2 },
        { text: 'Quase todos os dias', value: 3 }
      ]
    }
  },
  'WHO-5': {
    id: 'WHO-5',
    en: {
      name: 'WHO-5: Well-Being Index',
      description: 'A brief assessment of current subjective well-being.',
      questions: [
        'Have you been feeling cheerful, well, and in good spirits lately?',
        'How often have you felt calm, peaceful, and relaxed?',
        'Have you been feeling energetic, active, and full of vigor to carry out your tasks?',
        'When you wake up, do you feel that you had refreshing sleep and that you are truly rested?',
        'Do you feel that your day is filled with things and activities that spark your interest?'
      ],
      options: [
        { text: 'All the time', value: 5 },
        { text: 'Most of the time', value: 4 },
        { text: 'More than half the time', value: 3 },
        { text: 'Less than half the time', value: 2 },
        { text: 'Some of the time', value: 1 },
        { text: 'Never', value: 0 }
      ]
    },
    pt: {
      name: 'WHO-5: Índice de Bem-Estar',
      description: 'Uma avaliação breve do bem-estar subjetivo atual.',
      questions: [
        'Tem-se sentido alegre, bem e de bom humor ultimamente?',
        'Com que frequência se sentiu calmo, pacífico e relaxado?',
        'Tem-se sentido enérgico, ativo e cheio de vigor para realizar as suas tarefas?',
        'Quando acorda, sente que teve um sono reparador e que está verdadeiramente descansado?',
        'Sente que o seu dia está cheio de coisas e atividades que despertam o seu interesse?'
      ],
      options: [
        { text: 'O tempo todo', value: 5 },
        { text: 'A maior parte do tempo', value: 4 },
        { text: 'Mais da metade do tempo', value: 3 },
        { text: 'Menos da metade do tempo', value: 2 },
        { text: 'Parte do tempo', value: 1 },
        { text: 'Nunca', value: 0 }
      ]
    }
  },
  'PSS-10': {
    id: 'PSS-10',
    en: {
      name: 'PSS-10: Perceived Stress Scale',
      description: 'A brief assessment of perceived stress over the past weeks.',
      questions: [
        'Have you been affected or shaken by unexpected things that happened recently?',
        'How often have you felt that you had no control over things that really mattered in your life?',
        'Have you been feeling very nervous or "under pressure" lately?',
        'How often have you felt confident in your own ability to handle your personal problems?',
        'Have you been feeling that things are going the way you\'d like them to or as you planned?',
        'How often have you felt that you couldn\'t cope with all the tasks and responsibilities you had to do?',
        'Have you been able to control your annoyances and minor frustrations in your day-to-day life?',
        'How often have you felt that you had everything under control and that things were going well?',
        'Have you felt furious or very irritated because of things that were completely beyond your control?',
        'How often have you felt that difficulties were piling up so much that you couldn\'t get over them?'
      ],
      options: [
        { text: 'Never', value: 0 },
        { text: 'Almost never', value: 1 },
        { text: 'Sometimes', value: 2 },
        { text: 'Often', value: 3 },
        { text: 'Very often', value: 4 }
      ]
    },
    pt: {
      name: 'PSS-10: Escala de Stress Percebido',
      description: 'Uma avaliação breve do stress percecionado nas últimas semanas.',
      questions: [
        'Tem sido afetado ou abalado por coisas inesperadas que aconteceram recentemente?',
        'Com que frequência sentiu que não tinha controlo sobre o que realmente importa na sua vida?',
        'Tem-se sentido muito nervoso ou "sob pressão" ultimamente?',
        'Com que frequência se sentiu confiante na sua capacidade de lidar com os seus problemas pessoais?',
        'Tem sentido que as coisas estão a correr como gostaria ou como planeou?',
        'Com que frequência sentiu que não conseguia lidar com todas as tarefas e responsabilidades que tinha a fazer?',
        'Tem sido capaz de controlar as suas irritações e pequenas frustrações do dia-a-dia?',
        'Com que frequência sentiu que tinha tudo sob controlo e que as coisas corriam bem?',
        'Sentiu-se furioso ou muito irritado devido a coisas que estavam completamente fora do seu controlo?',
        'Com que frequência sentiu que as dificuldades se acumulavam tanto que não conseguia superá-las?'
      ],
      options: [
        { text: 'Nunca', value: 0 },
        { text: 'Quase nunca', value: 1 },
        { text: 'Às vezes', value: 2 },
        { text: 'Frequentemente', value: 3 },
        { text: 'Muito frequentemente', value: 4 }
      ]
    }
  }
};

const buildLocalizedObj = (qId, lang = 'pt') => {
  const q = QUESTIONNAIRES[qId];
  const langData = q[lang] || q['en'];
  return { id: q.id, ...langData };
};

const listQuestionnaires = (lang = 'pt') =>
  Object.keys(QUESTIONNAIRES).map(id => {
    const { name, description } = buildLocalizedObj(id, lang);
    return { id, name, description };
  });

const getQuestionnaire = (id, lang = 'pt') => {
  if (!id) return null;
  const qId = id.toUpperCase();
  if (!QUESTIONNAIRES[qId]) return null;
  return buildLocalizedObj(qId, lang);
};

export { QUESTIONNAIRES, listQuestionnaires, getQuestionnaire };
